### ISSUE TITLE
Improve manual labelling for depth sensors (simulation not working well)

### ISSUE THREAD

**Issue by danifpdra** [lardemua/atom#612]

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

can you describe the problem better?

**Comment by miguelriemoliveira** [comment]
Steps to reproduce:

    roslaunch rgb_depth_system_calibration dataset_playback.launch

then 

    clear && rosrun atom_calibration dataset_playback -json $ATOM_DATASETS/rgb_depth_system/rgb_depth_system_example_train_dataset/dataset.json -ow

and we get this labeling of the first dataset:

![image](https://github.com/lardemua/atom/assets/11350998/50534eb6-ab1b-462a-94fe-e0dcffacc740)

**Comment by miguelriemoliveira** [comment]
Tagging @brunofavs and @Kazadhum for visibility.

**Comment by miguelriemoliveira** [comment]
So I went through the tutorial video for [correcting depth labels](https://lardemua.github.io/atom_documentation/multimedia/#correcting-depth-labels), and the manual annotation is supposed to create lines right at the border of the pattern.

When I  annotate the polygon more carefully ... we get this annotation

![image](https://github.com/lardemua/atom/assets/11350998/b06fa3e0-0205-45be-9b61-e77407e338a3)

Which is much better. I had to be careful to make I did not have annotations away from the pattern (for this the 3d view is much better), and when that occurred I annotated again some 5 times until I got it right. 

This is what we have working before. It is not perfect since it requires a lot of manual work, but this is what we had.

I am thinking of ways to improve as per your request ... I will try.

**Comment by miguelriemoliveira** [comment]
So if I do a wrong annotation polygon, moving away from the pattern, See the red polygon (its hard to see, zoom in):

![image](https://github.com/lardemua/atom/assets/11350998/b512800a-0788-404a-be87-978271d25b58)

with this polygon , I get this:

![image](https://github.com/lardemua/atom/assets/11350998/b57f8298-6e2a-4ead-8027-a9d8cafd356f)

What would you want in this case @danifpdra ? Do you was that the polygon is well generated even with this bad annotarion of the polygon?

I think I could somehow filter out these very high jumps automatically. I am just not sure that would work in real data ...

**Comment by danifpdra** [comment]
> So if I do a wrong annotation polygon, moving away from the pattern, See the red polygon (its hard to see, zoom in):
> 
> ![image](https://user-images.githubusercontent.com/11350998/274764042-b512800a-0788-404a-be87-978271d25b58.png)
> 
> with this polygon , I get this:
> 
> ![image](https://user-images.githubusercontent.com/11350998/274763872-b57f8298-6e2a-4ead-8027-a9d8cafd356f.png)
> 
> What would you want in this case @danifpdra ? Do you was that the polygon is well generated even with this bad annotarion of the polygon?
> 
> I think I could somehow filter out these very high jumps automatically. I am just not sure that would work in real data ...

No, that's not the problem. The problem is that the jump from the ground to the "sky" in the simulation is not well perceived by the labeling and many times it is very hard to label because you have to get the corners absolutely right or it will not be well labeled. the general problem is not the pattern itself, but the borders.

**Comment by danifpdra** [comment]
Look at this example: 

![image](https://github.com/lardemua/atom/assets/47574905/42fadd44-7852-4273-9f4f-082164c96117)

![image](https://github.com/lardemua/atom/assets/47574905/9f5314db-bf69-40d2-b845-b29d18652c77)

Although I am clicking in places where there shouldn't be any doubts about where there is pattern and where there is not, the system is unable to understand that and assumes wherever I click, there is a border of the pattern and it is not able to perceive it's actual borders. Note that this does **not** happen in the real images. This is a **simulation** problem.

**Comment by miguelriemoliveira** [comment]
Thanks @danifpdra ,

I have an idea I will try to implement. But the part where you say this is a simulation only problem is what I do not understand. 

Can you give me steps to reproduce in real data?

**Comment by danifpdra** [comment]
hi @miguelriemoliveira,

Just follow these steps: https://github.com/lardemua/atom/issues/645#issue-1941950885

**Comment by miguelriemoliveira** [comment]
I am advancing (slowly). I have an idea and implemented it, to avoid getting those points very far away as boundaries.

I can create a mask which I give to function **labelDepthMsg**

https://github.com/lardemua/atom/blob/ab80d9aab2ddf5835e005f40df5c7db671c98623/atom_calibration/src/atom_calibration/dataset_playback/depth_manual_labeling.py#L76-L81

The problem now is that the function is returning boundary points outside that mask, which does not make sense ...

![image](https://github.com/lardemua/atom/assets/11350998/1690d0a5-a778-484f-ba84-7e804adf8e9a)

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

I think I made some progress. The labeling is working much better now.

Here's an example. I defined this polygon, not really worrying about being inside the pattern. Look at the left side of the pattern / polygon.

![image](https://github.com/lardemua/atom/assets/11350998/35a86f1a-84b6-4750-8707-35a0ff66a7f1)

And the result is:

![image](https://github.com/lardemua/atom/assets/11350998/904c89c3-ea4d-4e12-8c2e-c3a517b74d22)


Also for the boundaries I think it works. With this polygon:

![image](https://github.com/lardemua/atom/assets/11350998/598f7d84-2aa0-4012-a6b5-dfcc423e11af)

we get this segmentation:

![image](https://github.com/lardemua/atom/assets/11350998/a2bbc408-6862-4665-ae32-b253eff6bfe8)

Notice how there are no boundary magenta points at the bottom. 
Another example, with this user defined polygon:

![image](https://github.com/lardemua/atom/assets/11350998/ab2ec3ac-69c0-45b8-925d-2533ee0cc034)

we get:

![image](https://github.com/lardemua/atom/assets/11350998/d36c6f27-2ab4-4119-9f92-5615638a4ea0)

The downside is that this takes some 5 to 15 secs per segmentation. But in this case might as well wait and get a perfect result, that not wait and get unusable labels, you agree?

I only tested for simulation and using in the dataset playback. We need to test:

1. Simulation, during data collection, since there is a common function **labelDepthMsg** used in both
2. Test real data during data collection
3. Test real data during playback

Can you try these tests to see if / where it crashes? If I have time I will try them myself and let you know I did.

I am pushing the code to the main branch since you said the previous version was unusable.

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

I already tried case 1. There were some bugs I corrected.

Please try the simulation as well, and then give me some feedback on how this behaves with real data.

**Comment by miguelriemoliveira** [comment]
I am now trying with real data. We have some problems, namely we get this:

![image](https://github.com/lardemua/atom/assets/11350998/c1599296-988b-452b-8cc8-7fbd99c317af)

This is a really bad depth image ...

**Comment by miguelriemoliveira** [comment]
Another example

![image](https://github.com/lardemua/atom/assets/11350998/c8cf6f9d-ac34-4d88-beea-0decea8397b6)

![image](https://github.com/lardemua/atom/assets/11350998/b92cdefd-c1f8-46e6-8502-6bb311f04851)

**Comment by miguelriemoliveira** [comment]
And another.

![image](https://github.com/lardemua/atom/assets/11350998/46c19915-96a6-4a20-9824-318fa99c7f19)

It seems the new algorithm does not like the holes in the pattern which occur for real data.

The fact that boundaries are still there in the real data will be tackled in #644

**Comment by danifpdra** [comment]
Hummmm, that seems perfect for simulated data but real data is bound to have a lot of holes in it and that definitely doen't work, unless we create that manual eraser features for boundary points

**Comment by miguelriemoliveira** [comment]
We did create a manual eraser useful for real data (#648), so I am closing this one.
